src/host-config/proxy.ts

```diff
@@ -0,0 +1,270 @@
+import { promises as fs } from 'fs';
+import path from 'path';
+import { isRight } from 'fp-ts/lib/Either';
+import Reporter from 'io-ts-reporters';
+
+import type { RedsocksConfig, HostProxyConfig } from './types';
+import { ProxyConfig } from './types';
+import { pathOnBoot, readFromBoot, writeToBoot } from '../lib/host-utils';
+import { unlinkAll, mkdirp } from '../lib/fs-utils';
+import { isENOENT } from '../lib/errors';
+import log from '../lib/supervisor-console';
+import * as dbus from '../lib/dbus';
+
+const proxyBasePath = pathOnBoot('system-proxy');
+const noProxyPath = path.join(proxyBasePath, 'no_proxy');
+const redsocksConfPath = path.join(proxyBasePath, 'redsocks.conf');
+
+export const disallowedProxyFields = ['local_ip', 'local_port'];
+export const validRedsocksProxyTypes = [
+	'socks4',
+	'socks5',
+	'http-connect',
+	'http-relay',
+];
+
+const isAuthField = (field: string): boolean =>
+	['login', 'password'].includes(field);
+
+// ? is a lazy operator, so only the contents up until the first `}(?=\s|$)` is matched.
+// (?=\s|$) indicates that `}` must be followed by a whitespace or end of file to match,
+// in case there are user fields with brackets such as login or password fields.
+const blockRegexFor = (blockLabel: string) =>
+	new RegExp(`${blockLabel}\\s?{([\\s\\S]+?)}(?=\\s|$)`);
+
+const baseBlock = {
+	log_debug: 'off',
+	log_info: 'on',
+	log: 'stderr',
+	daemon: 'off',
+	redirector: 'iptables',
+};
+
+export class RedsocksConf {
+	public static stringify(config: RedsocksConfig): string {
+		const blocks: string[] = [];
+
+		if (config.redsocks) {
+			blocks.push(RedsocksConf.stringifyBlock('base', baseBlock));
+			blocks.push(
+				RedsocksConf.stringifyBlock('redsocks', config.redsocks, {
+					local_ip: '127.0.0.1',
+					local_port: 12345,
+				}),
+			);
+		}
+
+		return blocks.length ? blocks.join('\n') : '';
+	}
+
+	public static parse(rawConf: string): RedsocksConfig {
+		const conf: RedsocksConfig = {};
+		if (rawConf.length === 0) {
+			return conf;
+		}
+
+		// Extract contents of `redsocks {...}` using regex
+		const rawRedsocksBlockMatch = rawConf.match(blockRegexFor('redsocks'));
+		// If match length shorter than 2, no group was captured, indicating malformed config
+		if (!rawRedsocksBlockMatch || rawRedsocksBlockMatch.length < 2) {
+			log.warn('Invalid redsocks block in redsocks.conf');
+			return conf;
+		}
+		const rawRedsocksBlock = RedsocksConf.parseBlock(
+			rawRedsocksBlockMatch[1],
+			disallowedProxyFields,
+		);
+		const maybeProxyConfig = ProxyConfig.decode(rawRedsocksBlock);
+		if (isRight(maybeProxyConfig)) {
+			conf.redsocks = {
+				...maybeProxyConfig.right,
+			};
+		} else {
+			log.warn(Reporter.report(maybeProxyConfig).join('\n'));
+			// Allow the invalid values through for backwards compatibility
+			if (Object.keys(rawRedsocksBlock).length) {
+				conf.redsocks = rawRedsocksBlock as unknown as ProxyConfig;
```

Comment:
Returning the raw field might be too permissive as it will also return extra properties that would have been removed by the `decode`, I can agree that if the proxy type (for instance) is outside the supported values we can return it anyway (as long as we validate it on the input), but not sure we should return everything that is on the redsocks block.

Reply:
I was about to say that the Supervisor allowed invalid values through before when reading the redsocks block, but the current implementation actually also excludes invalid values. Modified this [here]() (TODO link)